karapace/compatibility.py

```diff
@@ -39,12 +46,226 @@ def check(self) -> bool:  # pylint: disable=inconsistent-return-statements
+        if isinstance(source, str):
+            source = {"type": source}
+        if isinstance(target, str):
+            target = {"type": target}
+        source_enum, target_enum = set(source.get("enum", [])), set(target.get("enum", []))
+        if len(target_enum.difference(source_enum)) > 0:
+            raise IncompatibleSchema(
+                f"Source enum values do {source_enum} do not fully include target enum values {target_enum}"
+            )
+        for format_key in ["pattern", "format"]:
+            source_pattern, target_pattern = source.get(format_key), target.get(format_key)
+            if source_pattern != target_pattern:
+                raise IncompatibleSchema(
+                    f"Source {format_key} {source_pattern} is different from target {format_key} {target_pattern}"
+                )
+        source_min_len, target_min_len = source.get("minLength", 0), target.get("minLength", 0)
+        if source_min_len > target_min_len:
+            raise IncompatibleSchema(f"Source min len {source_min_len} is not greater than target min len {target_min_len}")
+        source_max_len, target_max_len = source.get("maxLength", 0), target.get("maxLength", 0)
+        if source_max_len < target_max_len:
+            raise IncompatibleSchema(f"Source max len {source_max_len} is not greater than target max len {target_max_len}")
+        return True
+
+    def can_read_objects(self, source: dict, target: dict, checking_for: str) -> bool:
+        if source.get("propertyNames") != target.get("propertyNames"):
+            raise IncompatibleSchema(f"{source} and {target} have different values for propertyNames")
+        if source.get("minProperties", 0) > target.get("minProperties", 0):
+            raise IncompatibleSchema(f"{source} does not cover {target} on minProperties")
+        if source.get("maxProperties", 0) < target.get("maxProperties", 0):
+            raise IncompatibleSchema(f"{source} does not cover {target} on maxProperties")
+        source_additional_props = source.get("additionalProperties")
+        target_additional_props = target.get("additionalProperties")
+        if source_additional_props is False and target_additional_props is not False:
+            raise IncompatibleSchema(f"{source} does not cover {target} on additionalProperties")
+        if target_additional_props is True and source_additional_props is not True:
+            raise IncompatibleSchema(f"{source} does not cover {target} on additionalProperties")
+        if isinstance(source_additional_props, dict) and isinstance(target_additional_props, dict):
+            self.log.debug("Recursing for additional properties on %r and %r", source, target)
+            self.check_compatibility(source_additional_props, target_additional_props, checking_for)
+        return True
+        # no patternProperties, no dependencies
+
+    @staticmethod
+    def can_read_unions(source, target, checking_for: str):  # pylint: disable=unused-argument
```

Comment:
Is this supposed to be only returning True?

Reply:
yes, unions don't have properties to check for (like required fields or min / max value)